### ISSUE TITLE
CA1508 False Negative due to default typing of Literals

### ISSUE THREAD

**Issue by Applesauce314** [dotnet/roslyn-analyzers#6401]
### Analyzer
**Diagnostic ID**: [CA1508](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source
**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.303](https://dotnet.microsoft.com/download/dotnet/5.0)
**Version**: [SDK 6.0.402](https://dotnet.microsoft.com/download/dotnet/5.0)

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 5.0.3 
**Version**: 7.0.0

### Describe the bug

CA1508 can return false negatives when the dataflow analysis involves integer literals.

### Steps To Reproduce
Create code which involves numeric literals where the literal type does not match the type of a later statement which could be found to trigger CA1508.

These are obviously contrived examples, in a real scenario, the fact that there is a literal involved is often not apparent
##### This code has a CA1508 error
```csharp 
  public void Int()  
  {
    const int r = 10;
    if ((r + 50) == r)  
    {  }
  }
```
##### The exact same code with in changed to long does not have CA1508
```csharp 
  public void WithLong()  
  {
    const long r = 10;
    if ((r + 50) == r) 
    {  }
  }
```
##### If you declare the Literals as longs the error returns
```csharp with long the exact same code does not have CA1508
  public void WithLongLiterals()  
  {
    const long r = 10L;
    if ((r + 50) == r) 
    {  }
  }
```
I added unit tests to a [fork of the repository](https://github.com/Applesauce314/roslyn-analyzers/tree/CA1508-false-negative) that demonstrates the issue.

### Expected behavior
Value Content Analysis to not be stymied by literals.
### Actual behavior
Value Content Analysis is invalidated by the presence of literals that don't match type.
## Additional context
The point at which the error occurs is in these three code blocks. (The `TryMerge` function of [`ValueContentAbstractValue`](https://github.com/dotnet/roslyn-analyzers/blob/main/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/ValueContentAnalysis/ValueContentAbstractValue.cs#L287) and `TryConvertToUInt64` of [DiagnosticHelpers](https://github.com/dotnet/roslyn-analyzers/blob/60f592cb113cb6930300ed49b762f9c1e7abb849/src/Utilities/Compiler/DiagnosticHelpers.cs#L9) It assumes the `specialType` parameters passed in matches the actual type of the `value` parameter (or `value1`/`value2`), but when the type does not match the unboxing will throw an `InvalidCastException`, preventing the `ValueContentAnalysis` for the current statement from returning an analyzed value even though it should be able to.

https://github.com/dotnet/roslyn-analyzers/blob/60f592cb113cb6930300ed49b762f9c1e7abb849/src/Utilities/Compiler/DiagnosticHelpers.cs#L9-L58

https://github.com/dotnet/roslyn-analyzers/blob/60f592cb113cb6930300ed49b762f9c1e7abb849/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/ValueContentAnalysis/ValueContentAbstractValue.cs#L372

https://github.com/dotnet/roslyn-analyzers/blob/60f592cb113cb6930300ed49b762f9c1e7abb849/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/ValueContentAnalysis/ValueContentAbstractValue.cs#L298-L322

#### Notes
- The literal 0 will always have this issue even if declared as typed (0m, 0L, 0f, etc.) as 0 appears to have [special handling](https://github.com/dotnet/roslyn-analyzers/blob/60f592cb113cb6930300ed49b762f9c1e7abb849/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/ValueContentAnalysis/ValueContentAbstractValue.cs#L58). 
- an explicit cast does not help
- I don't know how this error interacts with [CS0162](https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0162)
-

**Comment by Applesauce314** [comment]
I created a [Proof of concept fix](https://github.com/Applesauce314/roslyn-analyzers/tree/CA1508-false-negative-fix) to demonstrate the problem can be fixed. It handles the cases I made tests for, but there are numerous more permutation the proof of concept does not handle. 
It appears the entire methodology these functions use to determine the type to unbox may need to be changed. 

- why is specialType used for the switch rather than the underlying type, is it performance? is the performance difference so bad a hybrid approach is needed? (e.g. first switch on specialType, then only use type checking if specialType is wrong?). 
- does fixing this possibly introduce false positives to this error that were masked by the failed casts?
- Could the value be corrected when the `LiteralValue` is created, instead of handling the mismatch later? (I think the operation needs to be fixed on the merge side, as the point at which the `LiteralValue` is created is only looking at the syntax of that literal.



**Addendum**: 
It is unfortunate that we are "stuck" in netstandard2.0 because the code that has problems is a result of the type of problem that led to the creation of the generic math features.

**Comment by Applesauce314** [comment]
I have been doing some digging on this guy and have made some inroads on the actual problem.
It appears that when `Visit` is called for an `IConversionOperation` during the analysis, it is returning the `ValueContentAbstractValue` with the literal's type equal to the before side of the conversion. Unfortunately, when I update `VisitConversion` to return _correctly_ typed literal values, it causes false positives in other unit tests. I will write up the result of my findings in more detail when I get the chance.
